### PR TITLE
resource limits for pods

### PR DESCRIPTION
**Pull request by cdoern**
added the following flags and handling for podman pod create

--memory-swap
--cpuset-mems
--device-read-bps
--device-write-bps
--blkio-weight
--blkio-weight-device
--cpu-shares

given the new backend for systemd in c/common, all of these can now be exposed to pod create.
most of the heavy lifting (nearly all) is done within c/common. However, some rewiring needed to be done here
as well!

Signed-off-by: Charlie Doern <cdoern@redhat.com>

#### Does this PR introduce a user-facing change?

```release-note
expose resource limts flags: 

--memory-swap
--cpuset-mems
--device-read-bps
--device-write-bps
--blkio-weight
--blkio-weight-device 
--cpu-shares

to podman pod create
```

### CODE REVIEW

**Comment by cdoern** [comment]
@giuseppe I am getting failures for `blkio-weight` and `blkio-weight-device` due to crun not being able to write `io.weight`. Looking around, it seems that crun might only be able to write to `io.bfq.weight`.

This might be something we want to look into... but for next week!

**Comment by giuseppe** [comment]
> @giuseppe I am getting failures for `blkio-weight` and `blkio-weight-device` due to crun not being able to write `io.weight`. Looking around, it seems that crun might only be able to write to `io.bfq.weight`.

we might need to add a fallback to `io.weight` when `io.bfq.weight` doesn't work.  I see in on Fedora though, what system are you using?

**Comment by giuseppe** [comment]
> we might need to add a fallback to `io.weight` when `io.bfq.weight` doesn't work. I see in on Fedora though, what system are you using?

opened a PR: https://github.com/containers/crun/pull/964

**Comment by cdoern** [comment]
@giuseppe how does getting a new crun version work since it isn't a go vendor? do we have to wait until a new release is cut or is it just built from the main branch.

**Comment by rhatdan** [comment]
@cdoern @giuseppe we need to get a release and then update the images we are testing with to use the new crun.

File: test/system/200-pod.bats
Comment (edsantiago) [inline test/system/200-pod.bats:487]: No no no. This is vandalism: not only does it corrupt an existing system, it prevents the test from ever running again because there's no cleanup (and even if there was cleanup, it would not run if there was an error). We do not vandalize our test system.

File: test/system/200-pod.bats
Comment (edsantiago) [inline test/system/200-pod.bats:488]: There really isn't a need for all the numberized variables. Please use "cname" or "containername", without creating unique variables for each one. Same applies to all the other numberized variables.

File: test/system/200-pod.bats
Comment (edsantiago) [inline test/system/200-pod.bats:496]: protip: error messages should always be unique. If/when this fails, the poor investigator will have no idea *what* resource failed. Even better, since this is growing so big, this is a good candidate for using `parse_table` and doing the check in a loop:
```bash
    run_podman pod inspect --format '{{.CgroupPath}}' $name1
    local cgroup_path="$output"

    expected_limits="
cpu.max         | 500000 100000
memory.max      | 5242880
memory.swap.max | 1068498944
io.max          | 1:3 rbps=1048576 wbps=1048576 riops=max wiops=max
"

    while read unit expect; do
        local actual=$(< /sys/fs/cgroup/$cgroup_path/$unit)
        is "$actual" "$expect" "resource limit: $unit"
    done < <(parse_table "$expected_limits")
    run_podman pod --cgroup-manager=systemd rm -f $name1
```

**Review by edsantiago — CHANGES_REQUESTED** [review]

File: test/system/200-pod.bats
Comment (cdoern) [reply test/system/200-pod.bats:487]: ok @edsantiago thanks for the feedback. I always forget that system tests have these consequences. I will mimic something more like the loopback device in 180-blkio.bats

File: test/system/200-pod.bats
Comment (edsantiago) [inline test/system/200-pod.bats:536]: Thank you, this is all much more readable now, but the problem with making things readable is that one discovers new problems :-)

As best I can tell, lines 527-536 are 100% identical to 515-524, with the only difference being `--cgroup-manager`. (And `--name`, which is irrelevant). Could you rewrite this as
```bash
    for cgm in systemd cgroupfs; do
        name=resources-$cgm
        run_podman --cgroup-manager=$cgm ...
        ....
        is "$actual" "$expect" "resource limit under $cgm: $unit"
```

**Comment by giuseppe** [comment]
one issue I've noticed is that you've hardcoded `/dev/block` for the device name.  But I think you should also consider `/dev/char` when the device is a character device.

I am not sure there is any advantage in resolving them, can't we just pass whatever name we go as input?

**Comment by TomSweeneyRedHat** [comment]
@cdoern all kinds or red tests here

**Comment by cdoern** [comment]
@giuseppe given the feedback in systemd/systemd#23938 I think we need to hardcore either block or char. Is there a way to tell which is which?

**Comment by rhatdan** [comment]
Drive by, can't you just stat the source and get whether it is a block or char device?

**Comment by rhatdan** [comment]
Major and minor are just ints, there is no way to differenctiate block or char.

**Comment by cdoern** [comment]
@giuseppe once the containers/common#1092 merges, I think all of these work for systemd and cgroupfs except for `--blkio-weight-device` only works for cgroupfs

**Comment by giuseppe** [comment]
looks like bfq is not enabled in your kernel.

What kernel are you using?

In any case, upstream I've added a fallback for crun to use `io.weight` when `io.bfq.weight` is not supported

**Comment by giuseppe** [comment]
no let's not worry about it, next crun will have the fix

**Comment by cdoern** [comment]
vendor seems large mostly due to the error switching in common. I think there is another podman PR vendoring in c/common but if not, we can just use this.

**Comment by cdoern** [comment]
agh getting the same error in the docs as the vendor treadmill PR.. will have to wait until that is resolved.

**Review by edsantiago — APPROVED** [review]

**Review by giuseppe — APPROVED** [review]
I am fine if we address the cpu-weight issue later; it is not so important anyway.

LGTM

**Comment by cdoern** [comment]
> I am fine if we address the cpu-weight issue later; it is not so important anyway.
> 
> LGTM

@giuseppe I made a runc issue for blkio-weight-device, seems to be an issue on their end. I think it is ok to merge this as well. see opencontainers/runc#3529

File: test/system/200-pod.bats
Comment (edsantiago) [inline test/system/200-pod.bats:526]: ...please do the `losetup -d` cleanup here too.

File: test/system/200-pod.bats
Comment (edsantiago) [inline test/system/200-pod.bats:489]: You could just do `LOOPDEVICE=$(losetup --show -f $lofile)` instead of your complicated loop below. `LOOPDEVICE` can be a global, defined at top (to null), and then checked in `teardown`.

File: test/system/200-pod.bats
Comment (edsantiago) [inline test/system/200-pod.bats:491]: ...then just do `losetup -l --noheadings --output MAJ:MIN $LOOPDEVICE`

File: test/system/200-pod.bats
Comment (edsantiago) [inline test/system/200-pod.bats:502]: Please do cleanup here

**Review by edsantiago — CHANGES_REQUESTED** [review]
Cleanup needed in aisle t... I mean, in tests. Sorry to be so nitpicky.

**Comment by cdoern** [comment]
@edsantiago let me know if I addressed all of your changes!

**Comment by cdoern** [comment]
@mheon @rhatdan @giuseppe no rush since this won't get into 4.2, but feel free to merge once we get @edsantiago 's blessing on the tests

**Comment by edsantiago** [comment]
No, tests are kind of broken still. Try this. 
[foo.diff.txt](https://github.com/containers/podman/files/9154276/foo.diff.txt)

**Comment by cdoern** [comment]
> No, tests are kind of broken still. Try this. [foo.diff.txt](https://github.com/containers/podman/files/9154276/foo.diff.txt)

that works.

**Comment by edsantiago** [comment]
K. Is it clear what I fixed? That is: is anything unclear? I'm sorry about providing a patch without explanation, it's just, argh, *today*!!

**Review by edsantiago — APPROVED** [review]

**Comment by cdoern** [comment]
> K. Is it clear what I fixed? That is: is anything unclear? I'm sorry about providing a patch without explanation, it's just, argh, *today*!!

No worries! I get everything you changed, I was trying to use sed for the spacing but yours works better.

**Comment by edsantiago** [comment]
LGTM but I have a feeling this is going to blow up when combined with #14972 (runc). If yours merges first, I will need your help getting tests to pass (I will probably just add yet more `Skip`s). If mine merges first, you will need to rebase this and then figure out the required `Skip`s or fixes.

**Comment by cdoern** [comment]
@mheon just an FYI for when you review, I just changed pod_api.go (pod inspect) to include the new limits since I forgot to do that. Also, @edsantiago skipped the resources limits test in v1 since the limits are not allowed in most cases.

File: libpod/options.go
Comment (mheon) [inline libpod/options.go:2148]: Can you throw a comment on here, just to make clear what this does?

**Comment by mheon** [comment]
Please add something to the `podman pod create` manpage about how resource limits for pods work. I think an explanation of the specifics (containers in the pod can have separate limits, the limit applies to all containers in the pod) and any exceptions (doesn't work if the container specifies its own cgroup) is worth it.

Code LGTM

**Comment by cdoern** [comment]
@mheon does this work?

``
Note: resource limit related flags work by setting the limits explicitly in the pod's cgroup
which by default, is the cgroup parent for all containers joining the pod. Containers are still delegated the ability to set their own resource limits when joining a pod via the **cgroup.subtree_control** file in the pod cgroup which impacts the **cgroup.controller** file in the child container cgroup. containers do NOT get the pod level cgroups if they specify their own cgroup when joining a pod such as **--cgroupns=host**
``

**Comment by mheon** [comment]
```ontainers are still delegated the ability to set their own resource limits when joining a pod via the **cgroup.subtree_control** file in the pod cgroup which impacts the **cgroup.controller** file in the child container cgroup```

This is very unclear. As a Podman user, what I care about is: Can I use `--cpu` on a pod to set an overall CPU limit, then use `--cpu` on a container in the pod to set a smaller limit for one container in the pod?

**Comment by cdoern** [comment]
> 

ok, @mheon 

``
Note: resource limit related flags work by setting the limits explicitly in the pod's cgroup
which by default, is the cgroup parent for all containers joining the pod. Containers are still delegated the ability to set their own resource limits when joining a pod meaning that if you execute **podman pod create --cpus=5** you can also execute **podman container create --pod=<POD_ID> --cpus=4** and the container will only see the smaller limit. containers do NOT get the pod level cgroup resources if they specify their own cgroup when joining a pod such as **--cgroupns=host**
``

**Comment by mheon** [comment]
That WFM

**Comment by cdoern** [comment]
@containers/podman-maintainers PTAL and merge

**Pull request merged**
